cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java

```diff
@@ -0,0 +1,106 @@
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.ByteBuffer;
+
+/**
+ * Download file from AppFabric via internal REST API calls.
+ * <p>
+ * The target file could be either residing in AppFabric's local file system or a distributed file system
+ * that AppFabric is configured to access.
+ */
+class FileFetcher {
+  private static final Logger LOG = LoggerFactory.getLogger(FileFetcher.class);
+  private RemoteClient remoteClient;
+
+  FileFetcher(DiscoveryServiceClient discoveryClient) {
+    this.remoteClient = new RemoteClient(discoveryClient,
+                                         Constants.Service.APP_FABRIC_HTTP,
+                                         new DefaultHttpRequestConfig(false),
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+  }
+
+  /**
+   * Download a file from AppFabric and store it in the target file.
+   *
+   * @param sourceURI uri to identity the file to download. This URI should exist in AppFabric.
+   * @param targetLocation target location to store the downloaded file
+   * @throws IOException if file downloading or writing to target location fails.
+   */
+  void download(URI sourceURI, Location targetLocation) throws IOException, BadRequestException {
+    OutputStream outputStream = targetLocation.getOutputStream();
+    HttpRequest request = remoteClient.requestBuilder(
+      HttpMethod.GET,
+      String.format("location/%s", sourceURI.getPath())).withContentConsumer(
+      new HttpContentConsumer() {
+        @Override
+        public boolean onReceived(ByteBuffer chunk) {
+          try {
+            byte[] bytes = new byte[chunk.remaining()];
+            chunk.get(bytes, 0, bytes.length);
+            outputStream.write(bytes);
+            outputStream.flush();
+          } catch (IOException e) {
+            LOG.error("Failed to download file from {} to {}", sourceURI, targetLocation.toURI(), e);
+            return false;
+          }
+          return true;
+        }
+
+        @Override
+        public void onFinished() {
+          try {
+            outputStream.close();
+          } catch (Exception e) {
+            LOG.error("Failed to close the file downloaded from {} to {}", sourceURI, targetLocation.toURI());
+          }
+        }
+      }).build();
+    HttpResponse httpResponse = HttpRequests.execute(request, new DefaultHttpRequestConfig(false));
+    httpResponse.consumeContent();
+
+    if (httpResponse.getResponseCode() != HttpResponseStatus.OK.code()) {
+      if (httpResponse.getResponseCode() == HttpResponseStatus.NOT_FOUND.code()) {
+        throw new FileNotFoundException(httpResponse.getResponseBodyAsString());
+      } else if (httpResponse.getResponseCode() == HttpResponseStatus.BAD_REQUEST.code()) {
```

Comment:
No need to have else after the if block throw. It is cleaner to write:

```java
if (...) {
  throw ...;
}
if (...) {
  throw ...
}
```

Reply:
Fixed.